src/ir_passes/compile.rs

```diff
@@ -0,0 +1,754 @@
+                    }),
+            )
+            .collect();
+
+        let mut interface_ports =
+            INTERFACE_PORTS.iter().collect::<HashSet<_>>();
+
+        // add interface port attributes if necessary
+        for pd in &mut ports {
+            if let Some(pair) = INTERFACE_PORTS
+                .iter()
+                .find(|(_, (n, _, _))| *n == pd.name.as_ref())
+            {
+                assert!(
+                    pair.1 .2 == pd.direction,
+                    "Expected {} to be an {:?} port, got {:?} port.",
+                    pair.1 .0,
+                    pair.1 .2,
+                    pd.direction
+                );
+                // TODO: Assert width equality
+                interface_ports.remove(pair);
+                pd.attributes.insert(pair.0 .0, pair.0 .1);
+            }
+        }
+
+        if is_comp {
+            // add remaining interface ports if not found
+            for (attr, (name, width, dir)) in interface_ports {
+                ports.push(calyx::PortDef {
+                    name: (*name).into(),
+                    width: wfu(*width),
+                    direction: dir.clone(),
+                    attributes: vec![*attr].try_into().unwrap(),
+                });
+            }
+        }
+
+        ports
+    }
+
+    /// Gets the component name given an [ir::Context] and an [ir::CompIdx]
+    fn comp_name(ctx: &ir::Context, idx: ir::CompIdx) -> String {
+        match &ctx.comps.get(idx).src_info {
+            Some(src_info) => src_info.name.to_string(),
+            None => format!("comp{}", idx.get()),
+        }
+    }
+
+    fn port_name(
+        ctx: &ir::Context,
+        comp: &ir::Component,
+        idx: ir::PortIdx,
+    ) -> String {
+        let p = comp.get(idx);
+
+        match &p.owner {
+            ir::PortOwner::Sig { .. } => {
+                if let Some(src) = &comp.src_info {
+                    src.ports.get(&idx).unwrap().to_string()
+                } else {
+                    format!("p{}", idx.get())
+                }
+            }
+            ir::PortOwner::Inv { base, .. } => {
+                base.apply(ctx, |c, p| Compile::port_name(ctx, c, p))
+            }
+            ir::PortOwner::Local => format!("p{}", idx.get()),
+        }
+    }
+
+    fn param_name(comp: &ir::Component, idx: ir::ParamIdx) -> String {
+        if let Some(src) = &comp.src_info {
+            src.params.get(&idx).unwrap().to_string()
+        } else {
+            format!("pr{}", idx.get())
+        }
+    }
+
+    // Gets the name associated with an event's interface port if it exists.
+    fn interface_port_name(
+        comp: &ir::Component,
+        idx: ir::EventIdx,
+    ) -> Option<String> {
+        if comp.get(idx).has_interface {
+            Some(if let Some(src) = &comp.src_info {
+                src.interface_ports.get(&idx).unwrap().to_string()
+            } else {
+                format!("ev{}", idx.get())
+            })
```

Comment:
This is a hard to read nesting of things. Use methods like `map` and `unwrap_or_else` on option to make it more readable